src/bootstrap/src/utils/build_stamp.rs

```diff
@@ -0,0 +1,194 @@
+//! Module for managing build stamp files.
+//!
+//! Contains the core implementation of how bootstrap utilizes stamp files on build processes.
+
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+use sha2::digest::Digest;
+
+use crate::core::builder::Builder;
+use crate::core::config::TargetSelection;
+use crate::utils::helpers::{hex_encode, mtime};
+use crate::{Compiler, Mode, helpers, t};
+
+#[cfg(test)]
+mod tests;
+
+/// Manages a stamp file to track build state. The file is created in the given
+/// directory and can have custom content and name.
+#[derive(Clone)]
+pub struct BuildStamp {
+    path: PathBuf,
+    pub(crate) stamp: String,
+}
+
+impl AsRef<Path> for BuildStamp {
+    fn as_ref(&self) -> &Path {
+        &self.path
+    }
+}
+
+impl BuildStamp {
+    /// Creates a new `BuildStamp` for a given directory.
+    ///
+    /// By default, stamp will be an empty file named `.stamp` within the specified directory.
+    pub fn new(dir: &Path) -> Self {
+        // Avoid using `is_dir()` as the directory may not exist yet.
+        // It is more appropriate to assert that the path is not a file.
+        assert!(!dir.is_file(), "can't be a file path");
+        Self { path: dir.join(".stamp"), stamp: String::new() }
+    }
+
+    /// Sets stamp content to the specified value.
+    pub fn with_stamp<S: ToString>(mut self, stamp: S) -> Self {
+        self.stamp = stamp.to_string();
+        self
+    }
+
+    /// Adds a prefix to stamp's name.
+    ///
+    /// Prefix cannot start or end with a dot (`.`).
+    pub fn with_prefix(mut self, prefix: &str) -> Self {
```

Comment:
Question [NON_LINEAR 2/2]: same here

Reply:
It is used multiple times - @klensy also pointed that earlier